pkg/registry/scripts/kube-client.js

```diff
@@ -0,0 +1,1033 @@
+            h |= 0; // Convert to 32bit integer
+        }
+        return Math.abs(h);
+    }
+
+    function search(arr, val) {
+        var low = 0;
+        var high = arr.length - 1;
+        var mid, v;
+
+        while (low <= high) {
+            mid = (low + high) / 2 | 0;
+            v = arr[mid];
+            if (v < val)
+                low = mid + 1;
+            else if (v > val)
+                high = mid - 1;
+            else
+                return mid; /* key found */
+        }
+        return low;
+    }
+
+    /**
+     * HashIndex
+     *
+     * A probablisting hash index, where items are added with
+     * various keys, and probable matches are returned. Similar
+     * to bloom filters, false positives are possible, but never
+     * false negatives.
+     */
+    function HashIndex(size) {
+        var self = this;
+        var array = [];
+
+        self.add = function add(keys, value) {
+            var i, j, p, x, length = keys.length;
+            for (j = 0; j < length; j++) {
+                i = hash("" + keys[j]) % size;
+                p = array[i];
+                if (p === undefined)
+                    p = array[i] = [];
+                x = search(p, value);
+                if (p[x] != value)
+                    p.splice(x, 0, value);
+            }
+        };
+
+        self.all = function all(keys) {
+            var i, j, p, result, n;
+            var rl, rv, pv, ri, px;
+
+            for (j = 0, n = keys.length; j < n; j++) {
+                i = hash("" + keys[j]) % size;
+                p = array[i];
+
+                /* No match for this key, short cut out */
+                if (!p) {
+                    result = [];
+                    break;
+                }
+
+                /* First key */
+                if (!result) {
+                    result = p.slice();
+
+                /* Calculate intersection */
+                } else {
+                    for (ri = 0, px = 0, rl = result.length; ri < rl; ) {
+                        rv = result[ri];
+                        pv = p[ri + px];
+                        if (pv < rv) {
+                            px += 1;
+                        } else if (rv !== pv) {
+                            result.splice(ri, 1);
+                            rl -= 1;
+                        } else {
+                            ri += 1;
+                        }
+                    }
+                }
+            }
+
+            return result || [];
+        };
+    }
+
+    function flat_schema(items) {
+        var i, len, ret = { "": DEFAULT };
+        for (i = 0, len = items.length; i < len; i++) {
+            ret[items[i].type] = items[i];
+            ret[items[i].kind] = items[i];
+        }
+        return ret;
+    }
+
+    /*
+     * Accepts:
+     *  1. an object
+     *  2. an involved object
+     *  2. a path string
+     *  3. type/kind, name, namespace
+     */
+    function resource_path(args) {
```

Comment:
Is there a case where it's called like this? I think all the callers pass an array of arguments.